### ISSUE TITLE
Problem with "PRAGMA journal_mode = WAL"

### ISSUE THREAD

**Issue by hanito** [aminroosta/sqlite_modern_cpp#4]
I am using the following commands in order to speed up the code.
database db(table.c_str());
db << "PRAGMA journal_mode = WAL";
db << "PRAGMA synchronous = NORMAL";

------------------------------------------------------------------------------------------

The problem is that I am getting the following error while debuging the code:
In __cxa_throw () (/usr/lib/x86_64-linux-gnu/libstdc++.so.6)
#1  0x000000000041e9e7 in sqlite::database_binder::~database_binder (this=0x7fffffffdb10, __in_chrg=<optimized out>) at/home/documents/3rdpartysqlite_modern_cpp.h:96
/home/documents/3rdparty/sqlite_modern_cpp.h:96:2480:beg:0x41e9e7
At /home/documents/3rdparty/sqlite_modern_cpp.h:96
Debugger finished with status 0

---------------------------------------------------------------------------------------------

Solution:

sqlite3 *handle;
sqlite3_open(table.c_str(), &handle);
sqlite3_exec(handle, "pragma journal_mode = WAL", NULL, NULL, NULL);
sqlite3_exec(handle, "PRAGMA synchronous = NORMAL", NULL, NULL, NULL);
database db(handle);
//WORK HERE
sqlite3_close(handle);

**Comment by Killili** [comment]
Interesting, the sqlite_step on that query returns "(100) SQLITE_ROW" and not "(101) SQLITE_DONE".
I think we should change line 95 to account for that.

int hresult = custom_sqlite3_step( _stmt );
if( hresult != SQLITE_DONE && hresult != SQLITE_ROW ) {....

**Comment by aminroosta** [comment]
From sqlite documentaion:  
*`Some pragmas take effect during the SQL compilation stage, not the execution stage. This means if using the C-language sqlite3_prepare(), sqlite3_step(), sqlite3_finalize() API (or similar in a wrapper interface), the pragma may run during the sqlite3_prepare() call, not during the sqlite3_step() call as normal SQL statements do. Or the pragma might run during sqlite3_step() just like normal SQL statements. Whether or not the pragma runs during sqlite3_prepare() or sqlite3_step() depends on the pragma and on the specific release of SQLite. `*

i see two options :  
**1**: **adding a member function `pragma` to the database class and using `sqlite3_exec`**.
**2**: **executing `sqlite3_step` multiple times in the database_binder destructor till `SQLITE_DONE` is returned**.

basically the second option is what `sqlite3_exec` actually does ...
if we go with the second option queries like this will run without errors:
```c++
   db << "select name from user"; // will no throw an exception
```

if we go with the first option ... then we have added a little bit of complexity to the library .
since the user should call different methods to do the same thing (right?).

both options are fine with me ... i have tested both of them ...

**Comment by Killili** [comment]
I would be for option 2, but it might back fire if we have a large query and don't want to run it till the end.
So option 1 it is for me, but lets call it "exec". 

But the question is, why is it not run in one step?

**Comment by aminroosta** [comment]
i don't know ... from what i've seen the first time i run `sqlite3_step` ,  `SQLITE_ROW` and the second time `SQLITE_DONE` is returned .   
the documentation is not clear :(

i wanted to force the user to always collect/handle the returned values from db.
that's the reason i throw an exception there .
(what do you mean by *` if we have a large query and don't want to run it till the end`*. ?

only **select** queries will be affected (**`db << "select col from table"`**).
i think that's not a big deal ... 

i'll commit the code ...

**Comment by aminroosta** [comment]
i've updated the code.

**Comment by Killili** [comment]
I was thinking of the example you gave, i as a user would not expect that all the rows of that query would be retrieved. But i admit that it would not make a lot of sense to have that query somewhere, other then a test case.

**Comment by hanito** [comment]
Hello Again sorry for the problems.  Can you help me please. Thank you
I created the following table:

----------------------------------------------------------------------------------------------------------
```
CREATE TABLE IF NOT EXISTS ENTITE_CANDIDATE (ID INTEGER PRIMARY KEY NOT NULL, ID_KBP TEXT NOT NULL, wiki_title TEXT, type TEXT NOT NULL, NOM TEXT NOT NULL, CLASSE TEXT, wikiPageInLinkCountCleaned TEXT, wikiPageRank TEXT, wikiHITS TEXT, wikiPageOutLinkCountCleaned TEXT, TFIDF TEXT);"
```
------------------------------------------------------------------------------------------------------------------------------
When I want to retrieve the informations by using the following query i get the following error:
```
db << "SELECT * FROM ENTITE_CANDIDATE WHERE ID =? ;"
                         << ids.first
                         >> [& ](int id, string id_kbp, string wiki_title, string type, string name, string classe, string wikiPageInLinkCountCleaned, string wikiPageRank, string wikiHITS, string wikiPageOutLinkCountCleaned, string tfidf)
                      {
                          cout <<" ==> " << tfidf << endl;
                          candidateEntities.push_back(make_tuple(id_kbp, type, tfidf));
                      };
```

```
||=== Build: Debug in entityLinking (compiler: GNU GCC Compiler) ===|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h||In instantiation of ‘sqlite::database_binder::operator>>(FUNC) [with FUNC = LinkEntities::link()::__lambda9]::__lambda6’:|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h|215|required from ‘struct sqlite::database_binder::operator>>(FUNC) [with FUNC = LinkEntities::link()::__lambda9]::__lambda6’|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h|214|required from ‘void sqlite::database_binder::operator>>(FUNC) [with FUNC = LinkEntities::link()::__lambda9]’|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/LinkEntities.cpp|86|required from here|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h|38|error: ‘static void sqlite::binder<N>::run(sqlite::database_binder&, F) [with F = LinkEntities::link()::__lambda9; int N = 11]’ is private|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h|215|error: within this context|
/home/hani/Documents/articles/Architecture/Modules/entityLinking/3rdparty/sqlite_modern_cpp.h|38|error: ‘static void sqlite::binder<N>::run(sqlite::database_binder&, F) [with F = LinkEntities::link()::__lambda9; int N = 11]’, declared using local type ‘LinkEntities::link()::__lambda9’, is used but never defined [-fpermissive]|
/usr/include/c++/4.8/functional|2443|error: ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = sqlite::database_binder::operator>>(FUNC) [with FUNC = LinkEntities::link()::__lambda9]::__lambda6; <template-parameter-2-2> = void; _Res = void; _ArgTypes = {}]’, declared using local type ‘sqlite::database_binder::operator>>(FUNC) [with FUNC = LinkEntities::link()::__lambda9]::__lambda6’, is used but never defined [-fpermissive]|
||=== Build failed: 4 error(s), 4 warning(s) (0 minute(s), 3 second(s)) ===|
```

**Comment by Killili** [comment]
Hello @hanito, 
i have opened a new issue for that.
